packages/signers/src/keypair-signer.ts

```diff
@@ -0,0 +1,64 @@
+import { Base58EncodedAddress, getAddressFromPublicKey } from '@solana/addresses';
+import { generateKeyPair, signBytes } from '@solana/keys';
+import { CompilableTransaction, ITransactionWithSignatures, signTransaction } from '@solana/transactions';
+
+import { MessageSigner, SignedMessageResponse, isMessageSigner } from './message-signer';
+import { TransactionSigner, isTransactionSigner } from './transaction-signer';
+
+/** Defines a signer capable of signing messages and transactions using a Crypto KeyPair. */
+export type KeyPairSigner<TAddress extends string = string> = MessageSigner<TAddress> &
+    TransactionSigner<TAddress> & { keyPair: CryptoKeyPair };
+
+/** Checks whether the provided value implements the {@link KeyPairSigner} interface. */
+export function isKeyPairSigner<TAddress extends string>(value: {
+    address: Base58EncodedAddress<TAddress>;
+}): value is KeyPairSigner<TAddress>;
+export function isKeyPairSigner(value: unknown): value is KeyPairSigner;
+export function isKeyPairSigner(value: unknown): value is KeyPairSigner {
+    return (
+        !!value &&
+        typeof value === 'object' &&
+        'keyPair' in value &&
+        typeof value.keyPair === 'object' &&
+        isMessageSigner(value) &&
+        isTransactionSigner(value)
+    );
+}
+
+/** Asserts that the provided value implements the {@link KeyPairSigner} interface. */
+export function assertIsKeyPairSigner<TAddress extends string>(value: {
+    address: Base58EncodedAddress<TAddress>;
+}): asserts value is KeyPairSigner<TAddress>;
+export function assertIsKeyPairSigner(value: unknown): asserts value is KeyPairSigner;
+export function assertIsKeyPairSigner(value: unknown): asserts value is KeyPairSigner {
```

Comment:
How come we have these two `unknown` ones? Won't everything valid meet the first overload with the `{address}` field? 
And nitpick: can we put the implementation on the first `value: unknown` one instead of writing it out twice?

Reply:
Oh that's because I though it might be helpful to have guard that accept any old piece of data and tell you if it's a signer or not. So technically you could do `assertIsKeyPairSigner("some_address_as_a_string")` and it will throw but type-check. That way if you have an input that's like `Based58EncodedAddress | KeyPairSigner` then you can thrown that input in there too. Would you prefer if the input guard was more narrow?

Reply:
I think I'd prefer to restrict it to types that might succeed, and make people narrow the type before calling it. I think that's what we've done with the other guards, eg you can't call `address(String | Number)`

Reply:
Gotcha, thanks! I'll update accordingly when I adjust the signer interfaces.